scripts/githooks/editorconfig-pre-commit.sh

```diff
@@ -1,32 +1,79 @@
 #!/bin/bash
 
-# Pre-commit git hook to check the EditorConfig rules complience over changed files
+set +e
+
+# Pre-commit git hook to check the EditorConfig rules compliance over changed
+# files. It ensures all non-binary files across the codebase are formatted
+# according to the style defined in the `.editorconfig` file.
 #
 # Usage:
 #   $ ./editorconfig-pre-commit.sh
 #
+# Options:
+#   BRANCH_NAME=other-branch-than-main  # Branch to compare with, default is `origin/main`
+#   ALL_FILES=true                      # Check all files, default is `false`
+#   VERBOSE=true                        # Show all the executed commands, default is `false`
+#
 # Exit codes:
-#   0 - all files are formatted correctly
-#   1 - files are not formatted correctly
+#   0 - All files are formatted correctly
+#   1 - Files are not formatted correctly
 #
 # Notes:
-#   Due to the file name escaping issue files are checked one by one.
+#   1) Please, make sure to enable EditorConfig linting in your IDE. For the
+#   Visual Studio Code editor it is `editorconfig.editorconfig` that is already
+#   specified in the `./.vscode/extensions.json` file.
+#   2) Due to the file name escaping issue files are checked one by one.
 
 # ==============================================================================
 
 exit_code=0
 image_digest=0f8f8dd4f393d29755bef2aef4391d37c34e358d676e9d66ce195359a9c72ef3 # 2.7.0
-changed_files=$(git diff --name-only --diff-filter=ACMRT)
 
-while read file; do
+# ==============================================================================
+
+function main() {
+
+  if is_arg_true "$ALL_FILES"; then
+
+    # Check all files
     docker run --rm --platform linux/amd64 \
-        --volume=$PWD:/check \
-        mstruebing/editorconfig-checker@sha256:$image_digest \
+      --volume=$PWD:/check \
+      mstruebing/editorconfig-checker@sha256:$image_digest \
+        ec \
+          --exclude '.git/'
+
+  else
+
+    # Check changed files only
+    changed_files=$(git diff --diff-filter=ACMRT --name-only ${BRANCH_NAME:-origin/main})
+    if [ -n "$changed_files" ]; then
+      while read file; do
+        docker run --rm --platform linux/amd64 \
+          --volume=$PWD:/check \
+          mstruebing/editorconfig-checker@sha256:$image_digest \
             ec \
-                --exclude '.git/' \
-                "$file"
-    [ $? != 0 ] && exit_code=1 ||:
-done < <(echo "$changed_files")
-[ $? != 0 ] && exit_code=1 ||:
+              --exclude '.git/' \
+              "$file"
+        [ $? != 0 ] && exit_code=1 ||:
+      done < <(echo "$changed_files")
+    fi
+
+  fi
```

Comment:
This all makes sense but I'm not too familiar with bash scripting syntax

Reply:
Good question. A number of things here -

1. `if [ -n "$changed_files" ]; then` we need to make sure that the changed files list is not empty otherwise editorconfig fails
2. `while read file; do ... done < <(echo "$changed_files")` the while loop passes the files one by one for linting. This is not ideal but it was the only way that I managed to make it working for files that contain spaces in their name. The issue here is that we pass the files (as a list or individually) for another execution inside Docker container. 
3. `[ $? != 0 ] && exit_code=1 ||:` this checks if the error code from `docker ...` is not `0` meaning it failed linting on a file. In that case, we set the `exit_code=1` variable for the whole script so it fails at the end but uses `||:` to swallow the error and continue linting the files so we can print the notification to the console informing which other files need to be looked after.